### PR TITLE
feat: add secret-key command line option

### LINKED ISSUE #1947: Deterministic Peer ID
### Describe the feature

It would be helpful to have the peer ID set deterministically. One way to do this is to set the secret key specifically, as done in #1946. Potentially there is a different API that would be warranted.

@gakonst @mattsse 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ryanorendorff**
fixes #1947 

The primary purpose of this PR is to be able to deterministically set the peer id, which is derived from the private key. The design here was to randomly generate the secret key if needed when using a CLI and then flatten the `Option<SecretKey>` so that downstream structs don't hold an `Option` value.

With a dummy private key of all `1`s:

```
2023-03-23T21:51:03.423309Z  INFO reth 0.1.0 starting
2023-03-23T21:51:03.424097Z  INFO Configuration loaded path=/Users/ryan/Library/Application Support/reth/db
2023-03-23T21:51:03.424114Z  INFO Opening database path=/Users/ryan/Library/Application Support/reth/db
2023-03-23T21:51:03.425595Z  INFO Database opened
2023-03-23T21:51:03.425634Z DEBUG Initializing genesis chain=mainnet genesis=0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
2023-03-23T21:51:03.667181Z  INFO Consensus engine initialized
2023-03-23T21:51:03.667248Z  INFO Test transaction pool initialized
2023-03-23T21:51:03.667259Z  INFO Connecting to P2P network
2023-03-23T21:51:03.672462Z  INFO Connected to P2P network peer_id=0x4f35…f1c1 local_addr=0.0.0.0:30303
2023-03-23T21:51:03.672487Z DEBUG Full peer id 0x4f355bdcb7cc0af728ef3cceb9615d90684bb5b2ca5f859ab0f0b704075871aa385b6b1b8ead809ca67454d9683fcf2ba03456d6fe2c4abe2b07f0fbdbb2f1c1
2023-03-23T21:51:03.672577Z  INFO Started RPC server
```

Running a second time produces the same peer id.

### CODE REVIEW

**Review by onbjerg — CHANGES_REQUESTED** [review]
I would prefer this to be a path so we don't leak secret keys on the command line, and we should also note in the code that it is temporary (we should write the secret key somewhere later on)

**Comment by ryanorendorff** [comment]
Sounds good! I can make that change.

**Comment by ryanorendorff** [comment]
@onbjerg I can also just add writing the secret key in `data_dir()/secret_key/secret` or some variant of that path in this PR if that makes sense. To match the `jwt` secret setup.

**Comment by gakonst** [comment]
Yeah let's write the key in `data_dir()/p2p/secret`

**Comment by ryanorendorff** [comment]
@gakonst Implemented the interface as specified! This will mean that each run of `reth node` will have the same peer ID after the initial secret key is created.

**Comment by ryanorendorff** [comment]
Huh the `test_session_established_with_different_capability` test passes for me locally, not sure what the difference is. 🤔

**Comment by gakonst** [comment]
Thanks - we're heads down sprinting for mainnet Shanghai hence unable to look at this deeply so far. If you keep encountering this issue, we'll take a closer look mid-next week! Ty for the patience :)

File: bin/reth/src/args/multi_use_args.rs
Comment (mattsse) [inline bin/reth/src/args/multi_use_args.rs:2]: I expect that we will get more "multi_use_arg" types.

but this file also doesn't define args, I'd like to rename this to `secret_key` instead

File: bin/reth/src/args/multi_use_args.rs
Comment (mattsse) [inline bin/reth/src/args/multi_use_args.rs:45]: is this effectively a `cp` operation?

File: bin/reth/src/dirs.rs
Comment (mattsse) [inline bin/reth/src/dirs.rs:62]: this is the secret key for p2p/network, I'd like to make this clear in the function name

File: bin/reth/src/p2p/mod.rs
Comment (mattsse) [inline bin/reth/src/p2p/mod.rs:48]: same here

File: bin/reth/src/stage/mod.rs
Comment (mattsse) [inline bin/reth/src/stage/mod.rs:65]: same

**Review by mattsse — CHANGES_REQUESTED** [review]

File: bin/reth/src/dirs.rs
Comment (ryanorendorff) [reply bin/reth/src/dirs.rs]: Done!

File: bin/reth/src/args/multi_use_args.rs
Comment (ryanorendorff) [reply bin/reth/src/args/multi_use_args.rs]: So this implements what @gakonst recommended:

1. `Ok(true)`: If the secret key exists, then it reads the file.
2. `Ok(false)`: If the secret key does not exist but the path is accessible, then create a new secret key and write it to disk.
3. `Err(e)`: if the path cannot be reached (invalid permissions, etc), error.

I think it is a tad different than `cp`; it saves the key to disk if it doesn't exist and uses an existing key if the key does exist.

One consequence of this PR is that the default behavior changes. After the first `reth` call, the p2p secret will always be the same on subsequent runs and similarly the node peer ID will be static. On the main branch, currently the p2p secret and node ID changes every time `reth` is run.

**Comment by ryanorendorff** [comment]
Made the changes. I renamed the `multi_use_args` to `secret_key`; if you have a preferred location for this file let me know. As mentioned, this code is not really an argument. I was just unsure where to store the code.

File: bin/reth/src/stage/mod.rs
Comment (mattsse) [inline bin/reth/src/stage/mod.rs:63]: It looks like Geth doesn't have this option, does it?

so we're free to choose a name here, imo p2p_ or network_ prefix is fine.

**Review by mattsse — APPROVED** [review]
code lgtm, pending @onbjerg 

but I'm unsure about the default behaviour re static peerid

File: bin/reth/src/args/multi_use_args.rs
Comment (mattsse) [reply bin/reth/src/args/multi_use_args.rs]: > After the first reth call, the p2p secret will always be the same on subsequent runs and similarly the node peer ID will be static.

hmm, i am a bit skeptical about this, imo this should always require opt-in.

File: bin/reth/src/stage/mod.rs
Comment (ryanorendorff) [reply bin/reth/src/stage/mod.rs:63]: It looks like `geth` does not have this option, although I am not super familiar with that execution client.

**Comment by ryanorendorff** [comment]
> but I'm unsure about the default behavior re static peer id

I can change it to do the current default behavior if you would like. This could be done by overwriting the file instead and make secret key loading require the key to exist on disk.

File: bin/reth/src/stage/mod.rs
Comment (ryanorendorff) [reply bin/reth/src/stage/mod.rs:63]: `geth` produces the same peer ID all the time because it uses a data directory. So the default behavior change seems to be closer to what `geth` does.

This does seem like potentially a `--datadir` like option would be the best in the end, which remaps all the default directories to the provided one.

Comment (mattsse) [reply bin/reth/src/stage/mod.rs:63]: ah okay, in that case this behaviour should be in line with expectations

ty

**Review by mattsse — APPROVED** [review]
lgtm, 
perhaps we should block merging this until we have https://github.com/paradigmxyz/reth/issues/1840
@gakonst ?

**Review by gakonst — APPROVED** [review]
OK on further thought let's merge this, sure we'll have the same identity across networks for a little bit, main downside. And that PR will fix that.

**Pull request merged**
